### PR TITLE
detect explicit sync and overlay usage in runtime

### PR DESCRIPTION
**Pull request by mwang30**
Jira: None
Test: pass test with AndroidIA and Android Things

Signed-off-by: Mingwei Wang <mingwei.wang@intel.com>

### CODE REVIEW

File: common/compositor/gl/egloffscreencontext.cpp [hunk 1]
```diff
@@ -102,22 +103,22 @@ void EGLOffScreenContext::RestoreState() {
...
   EGLint sync_fd = -1;
-#ifndef DISABLE_EXPLICIT_SYNC
-  EGLSyncKHR egl_sync =
-      eglCreateSyncKHR(egl_display_, EGL_SYNC_NATIVE_FENCE_ANDROID, NULL);
-  if (egl_sync == EGL_NO_SYNC_KHR) {
-    ETRACE("Failed to make sync object.");
-    return -1;
-  }
+  if (android::DrmHwcTwo::IsExplicitSyncEnabled()) {
+    EGLSyncKHR egl_sync =
```
Comment (kalyankondapally) [inline common/compositor/gl/egloffscreencontext.cpp]: white space ? common/ should not include any android specific or linux specific headers

File: common/display/displayplane.cpp [hunk 2]
```diff
@@ -126,13 +127,13 @@ bool DisplayPlane::Initialize(uint32_t gpu_fd,
...
     ETRACE("Could not get alpha property");
 
-#ifndef DISABLE_EXPLICIT_SYNC
-  ret = in_fence_fd_prop_.Initialize(gpu_fd, "IN_FENCE_FD", plane_props);
-  if (!ret)
-    ETRACE("Could not get IN_FENCE_FD property");
-#else
-  in_fence_fd_prop_.id = 0;
-#endif
+  if (android::DrmHwcTwo::IsExplicitSyncEnabled()) {
```
Comment (kalyankondapally) [inline common/display/displayplane.cpp]: white space ? common/ should not include any android specific or linux specific headers

File: common/compositor/gl/egloffscreencontext.cpp (hunk 1 above)
Comment (kalyankondapally) [inline common/compositor/gl/egloffscreencontext.cpp]: Instead of doing all this, what you can do is add a new API in https://github.com/01org/IA-Hardware-Composer/blob/master/public/nativedisplay.h void DisableExplicitSyncSupport()  and NativeDisplay will propogate it to compositor etc during creating time. DisableExplicitSyncSupport can be called from drmhwctwo during initialization phase depending on run time property.

File: common/compositor/gl/egloffscreencontext.h [hunk 3]
```diff
@@ -46,6 +47,7 @@ class EGLOffScreenContext {
   EGLSurface saved_egl_read_ = EGL_NO_SURFACE;
   EGLSurface saved_egl_draw_ = EGL_NO_SURFACE;
   bool restore_context_;
+  std::unique_ptr<OS> os_;
```
Comment (kalyankondapally) [inline common/compositor/gl/egloffscreencontext.h]: hmm we shouldnt need this. As I mentioned in last review add an API to NativeDisplay i.e. DisableExplicitSync and during displayinitialization in drmhwctwo, we can call that depending on runtime property.

File: os/android/drmhwctwo.cpp [hunk 4]
```diff
@@ -65,12 +84,15 @@ DrmHwcTwo::DrmHwcTwo() {
...
 
 HWC2::Error DrmHwcTwo::Init() {
+  OS::GetExplicitSync();
+
   displays_.emplace(std::piecewise_construct,
                     std::forward_as_tuple(HWC_DISPLAY_PRIMARY),
                     std::forward_as_tuple(&device_, HWC_DISPLAY_PRIMARY,
                                           HWC2::DisplayType::Physical));
 
   displays_.at(HWC_DISPLAY_PRIMARY).Init();
```
Comment (kalyankondapally) [inline os/android/drmhwctwo.cpp]: here in Init call you could query the run time setting and check if we need to disabled explicit sync for example

File: common/compositor/gl/egloffscreencontext.cpp [hunk 5]
```diff
@@ -20,6 +20,8 @@
 
 namespace hwcomposer {
 
+bool hwcomposer::EGLOffScreenContext::is_ExplicitSync_enabled_ = true;
```
Comment (kalyankondapally) [inline common/compositor/gl/egloffscreencontext.cpp]: can be a private member variable and doesnt have to be static

File: common/compositor/gl/egloffscreencontext.h [hunk 6]
```diff
@@ -38,6 +38,7 @@ class EGLOffScreenContext {
 
   void RestoreState();
 
+  static void SetExplicitSync(bool ExplicitSyncEnabled);
```
Comment (kalyankondapally) [inline common/compositor/gl/egloffscreencontext.h]: doesnt have to be static function.

File: common/compositor/gl/egloffscreencontext.h [hunk 7]
```diff
@@ -46,6 +47,7 @@ class EGLOffScreenContext {
   EGLSurface saved_egl_read_ = EGL_NO_SURFACE;
   EGLSurface saved_egl_draw_ = EGL_NO_SURFACE;
   bool restore_context_;
+  static bool is_ExplicitSync_enabled_;
```
Comment (kalyankondapally) [inline common/compositor/gl/egloffscreencontext.h]: doesnt have to be static

File: common/compositor/gl/glrenderer.cpp [hunk 8]
```diff
@@ -25,6 +25,8 @@
 
 namespace hwcomposer {
 
+bool hwcomposer::GLRenderer::is_ExplicitSync_enabled_ = true;
```
Comment (kalyankondapally) [inline common/compositor/gl/glrenderer.cpp]: can be a private member variable and doesnt have to be static

File: common/compositor/gl/glrenderer.h [hunk 9]
```diff
@@ -42,13 +42,15 @@ class GLRenderer : public Renderer {
 
   bool MakeCurrent() override;
 
+  static void SetExplicitSync(bool ExplicitSyncEnabled);
```
Comment (kalyankondapally) [inline common/compositor/gl/glrenderer.h]: doesnt have to be static

File: common/compositor/gl/glrenderer.h [hunk 10]
```diff
@@ -42,13 +42,15 @@ class GLRenderer : public Renderer {
...
  private:
   GLProgram *GetProgram(unsigned texture_count);
 
   EGLOffScreenContext context_;
 
   std::vector<std::unique_ptr<GLProgram>> programs_;
   GLuint vertex_array_ = 0;
+  static bool is_ExplicitSync_enabled_;
```
Comment (kalyankondapally) [inline common/compositor/gl/glrenderer.h]: doesnt have to be static

File: common/display/displayplane.cpp [hunk 11]
```diff
@@ -25,6 +25,8 @@
 
 namespace hwcomposer {
 
+bool hwcomposer::DisplayPlane::is_ExplicitSync_enabled_ = true;
```
Comment (kalyankondapally) [inline common/display/displayplane.cpp]: can be a private member variable and doesnt have to be static

File: common/display/displayplane.h [hunk 12]
```diff
@@ -96,6 +98,7 @@ class DisplayPlane {
   bool enabled_;
 
   std::vector<uint32_t> supported_formats_;
+  static bool is_ExplicitSync_enabled_;
```
Comment (kalyankondapally) [inline common/display/displayplane.h]: can be a private member variable and doesnt have to be static

File: common/display/displayplanemanager.cpp [hunk 13]
```diff
@@ -30,6 +30,8 @@
 
 namespace hwcomposer {
 
+bool hwcomposer::DisplayPlaneManager::is_ExplicitSync_enabled_ = true;
```
Comment (kalyankondapally) [inline common/display/displayplanemanager.cpp]: can be a private member variable and doesnt have to be static

File: common/display/displayplanemanager.h [hunk 14]
```diff
@@ -64,6 +64,10 @@ class DisplayPlaneManager {
   void EndFrameUpdate();
   bool CheckPlaneFormat(uint32_t format);
 
+  static void SetExplicitSync(bool ExplicitSyncEnabled);
+ private:
+  static bool is_ExplicitSync_enabled_;
```
Comment (kalyankondapally) [inline common/display/displayplanemanager.h]: can be a private member variable and doesnt have to be static

File: common/display/displayqueue.cpp [hunk 15]
```diff
@@ -29,6 +29,8 @@
 
 namespace hwcomposer {
 
+bool hwcomposer::DisplayQueue::is_ExplicitSync_enabled_ = true;
```
Comment (kalyankondapally) [inline common/display/displayqueue.cpp]: can be a private member variable and doesnt have to be static

File: common/display/displayqueue.h [hunk 16]
```diff
@@ -115,6 +116,7 @@ class DisplayQueue {
   std::vector<OverlayLayer> previous_layers_;
   DisplayPlaneStateList previous_plane_state_;
   OverlayBufferManager* buffer_manager_;
+  static bool is_ExplicitSync_enabled_;
```
Comment (kalyankondapally) [inline common/display/displayqueue.h]: can be a private member variable and doesnt have to be static

File: common/display/display.cpp [hunk 17]
```diff
@@ -244,4 +248,14 @@ void Display::SetBrightness(uint32_t red, uint32_t green, uint32_t blue) {
 bool Display::SetBroadcastRGB(const char *range_property) {
   return display_queue_->SetBroadcastRGB(range_property);
 }
+
+void Display::SetExplicitSync(bool ExplicitSyncEnabled) {
+  is_ExplicitSync_enabled_ = ExplicitSyncEnabled;
+  display_queue_->SetExplicitSync(ExplicitSyncEnabled);
```
Comment (kalyankondapally) [inline common/display/display.cpp]: here u can displayqueue like now and in displayqueue u can call displayplanemanager and compositor. DisplayPlaneManager calls into displayplane. Compositor calls into GLRenderer, dont need static functions

File: common/compositor/compositor.cpp [hunk 18]
```diff
@@ -40,19 +40,22 @@ Compositor::~Compositor() {
 }
 
 void Compositor::Init() {
-  gpu_resource_handler_.reset(CreateNativeGpuResourceHandler());
```
Comment (kalyankondapally) [inline common/compositor/compositor.cpp]: Can u separate these changes as they are not related

File: common/compositor/compositor.cpp [hunk 19]
```diff
@@ -264,4 +267,8 @@ void Compositor::SeparateLayers(const std::vector<size_t> &dedicated_layers,
   }
 }
 
+void Compositor::SetExplicitSync(bool ExplicitSyncEnabled) {
```
Comment (kalyankondapally) [inline common/compositor/compositor.cpp]: ExplicitSyncEnabled -> explicit_sync_enabled

File: common/compositor/compositor.h [hunk 20]
```diff
@@ -51,6 +51,7 @@ class Compositor {
                      uint32_t height, HWCNativeHandle output_handle,
                      int32_t *retire_fence);
   void InsertFence(uint64_t fence);
+  void SetExplicitSync(bool ExplicitSyncEnabled);
```
Comment (kalyankondapally) [inline common/compositor/compositor.h]: ExplicitSyncEnabled -> explicit_sync_enabled

File: common/compositor/gl/egloffscreencontext.cpp [hunk 21]
```diff
@@ -102,23 +102,27 @@ void EGLOffScreenContext::RestoreState() {
...
 EGLint EGLOffScreenContext::GetSyncFD() {
   EGLint sync_fd = -1;
-#ifndef DISABLE_EXPLICIT_SYNC
-  EGLSyncKHR egl_sync =
-      eglCreateSyncKHR(egl_display_, EGL_SYNC_NATIVE_FENCE_ANDROID, NULL);
-  if (egl_sync == EGL_NO_SYNC_KHR) {
-    ETRACE("Failed to make sync object.");
-    return -1;
-  }
+  if (is_ExplicitSync_enabled_) {
```
Comment (kalyankondapally) [inline common/compositor/gl/egloffscreencontext.cpp]: is_explicit_sync_enabled_

File: common/compositor/gl/egloffscreencontext.cpp [hunk 22]
```diff
@@ -102,23 +102,27 @@ void EGLOffScreenContext::RestoreState() {
...
+    }
 
-  eglDestroySyncKHR(egl_display_, egl_sync);
-#endif
+    eglDestroySyncKHR(egl_display_, egl_sync);
+  }
   return sync_fd;
 }
 
+void EGLOffScreenContext::SetExplicitSync(bool ExplicitSyncEnabled) {
```
Comment (kalyankondapally) [inline common/compositor/gl/egloffscreencontext.cpp]: ExplicitSyncEnabled -> explicit_sync_enabled

File: common/compositor/gl/glrenderer.cpp [hunk 23]
```diff
@@ -123,18 +123,18 @@ bool GLRenderer::MakeCurrent() {
...
-      EGL_NONE,
-  };
-  EGLSyncKHR fence = eglCreateSyncKHR(
-      context_.GetDisplay(), EGL_SYNC_NATIVE_FENCE_ANDROID, attrib_list);
-  eglWaitSyncKHR(context_.GetDisplay(), fence, 0);
-  eglDestroySyncKHR(context_.GetDisplay(), fence);
-#else
-  glFlush();
-#endif
+  if (is_ExplicitSync_enabled_) {
```
Comment (kalyankondapally) [inline common/compositor/gl/glrenderer.cpp]: is_explicit_sync_enabled_

File: common/compositor/gl/glrenderer.cpp [hunk 24]
```diff
@@ -156,4 +156,9 @@ GLProgram *GLRenderer::GetProgram(unsigned texture_count) {
   return 0;
 }
 
+void GLRenderer::SetExplicitSync(bool ExplicitSyncEnabled) {
```
Comment (kalyankondapally) [inline common/compositor/gl/glrenderer.cpp]: ExplicitSyncEnabled -> explicit_sync_enabled

File: common/compositor/gl/glrenderer.cpp [hunk 25]
```diff
@@ -156,4 +156,9 @@ GLProgram *GLRenderer::GetProgram(unsigned texture_count) {
...
+  is_ExplicitSync_enabled_ = ExplicitSyncEnabled;
```
Comment (kalyankondapally) [inline common/compositor/gl/glrenderer.cpp]: is_explicit_sync_enabled_

File: common/compositor/gl/glrenderer.h [hunk 26]
```diff
@@ -42,13 +42,15 @@ class GLRenderer : public Renderer {
 
   bool MakeCurrent() override;
 
+  void SetExplicitSync(bool ExplicitSyncEnabled) override;
```
Comment (kalyankondapally) [inline common/compositor/gl/glrenderer.h]: ExplicitSyncEnabled -> explicit_sync_enabled

File: common/compositor/gl/glrenderer.h [hunk 27]
```diff
@@ -42,13 +42,15 @@ class GLRenderer : public Renderer {
...
  private:
   GLProgram *GetProgram(unsigned texture_count);
 
   EGLOffScreenContext context_;
 
   std::vector<std::unique_ptr<GLProgram>> programs_;
   GLuint vertex_array_ = 0;
+  bool is_ExplicitSync_enabled_;
```
Comment (kalyankondapally) [inline common/compositor/gl/glrenderer.h]: is_explicit_sync_enabled_

File: common/compositor/renderer.h [hunk 28]
```diff
@@ -43,6 +43,8 @@ class Renderer {
   virtual void RestoreState() = 0;
 
   virtual bool MakeCurrent() = 0;
+
+  virtual void SetExplicitSync(bool ExplicitSyncEnabled) = 0;
```
Comment (kalyankondapally) [inline common/compositor/renderer.h]: ExplicitSyncEnabled -> explicit_sync_enabled

File: common/display/display.h [hunk 29]
```diff
@@ -94,6 +94,7 @@ class Display : public NativeDisplay {
   void SetContrast(uint32_t red, uint32_t green, uint32_t blue) override;
   void SetBrightness(uint32_t red, uint32_t green, uint32_t blue) override;
   bool SetBroadcastRGB(const char *range_property) override;
+  void SetExplicitSync(bool ExplicitSyncEnabled) override;
```
Comment (kalyankondapally) [inline common/display/display.h]: ExplicitSyncEnabled -> explicit_sync_enabled

File: common/display/displayplane.cpp [hunk 30]
```diff
@@ -126,13 +126,13 @@ bool DisplayPlane::Initialize(uint32_t gpu_fd,
...
     ETRACE("Could not get alpha property");
 
-#ifndef DISABLE_EXPLICIT_SYNC
-  ret = in_fence_fd_prop_.Initialize(gpu_fd, "IN_FENCE_FD", plane_props);
-  if (!ret)
-    ETRACE("Could not get IN_FENCE_FD property");
-#else
-  in_fence_fd_prop_.id = 0;
-#endif
+  if (DisplayPlane::is_ExplicitSync_enabled_) {
```
Comment (kalyankondapally) [inline common/display/displayplane.cpp]: ?

File: common/display/displayplane.h [hunk 31]
```diff
@@ -63,6 +63,8 @@ class DisplayPlane {
 
   void Dump() const;
 
+  void SetExplicitSync(bool ExplicitSyncEnabled);
```
Comment (kalyankondapally) [inline common/display/displayplane.h]: ExplicitSyncEnabled -> explicit_sync_enabled

File: common/display/displayplanemanager.cpp [hunk 32]
```diff
@@ -125,20 +125,20 @@ std::tuple<bool, DisplayPlaneStateList> DisplayPlaneManager::ValidateLayers(
...
-                         &composition, &render_layers);
-    if (!composition.empty())
-      return std::make_tuple(render_layers, std::move(composition));
-  }
-
-  // Dont use cache next frame if we are doing modeset now. With modeset
-  // we force all layers for 3D composition.
-  use_cache_ = !pending_modeset;
-#endif
+  if (is_ExplicitSync_enabled_) {
```
Comment (kalyankondapally) [inline common/display/displayplanemanager.cpp]: is_explicit_sync_enabled_

File: common/display/displayplanemanager.h [hunk 33]
```diff
@@ -64,6 +64,10 @@ class DisplayPlaneManager {
   void EndFrameUpdate();
   bool CheckPlaneFormat(uint32_t format);
 
+  void SetExplicitSync(bool ExplicitSyncEnabled);
```
Comment (kalyankondapally) [inline common/display/displayplanemanager.h]: ExplicitSyncEnabled -> explicit_sync_enabled

File: common/compositor/compositor.cpp (hunk 18 above)
Comment (mwang30) [inline common/compositor/compositor.cpp]: renderer_ need to be initialized in Compositor::Init(). If not, below call will caused fault address error.
renderer_->SetExplicitSync(ExplicitSyncEnabled);

File: os/android/drmhwctwo.h [hunk 34]
```diff
@@ -239,6 +239,9 @@ class DrmHwcTwo : public hwc2_device_t {
   hwcomposer::GpuDevice device_;
   std::map<hwc2_display_t, HwcDisplay> displays_;
   std::map<HWC2::Callback, HwcCallback> callbacks_;
+
+ private:
+  static bool is_explicit_sync_enabled_;
```
Comment (mwang30) [inline os/android/drmhwctwo.h]: If remove static, there is compile issue which hard to be work around:

vendor/intel/external/android_ia/hwcomposer/os/android/drmhwctwo.cpp:192:29: error: use of non-static data member 'is_explicit_sync_enabled_' of 'DrmHwcTwo' from nested type 'HwcDisplay'
  display_->SetExplicitSync(is_explicit_sync_enabled_);
                            ^~~~~~~~~~~~~~~~~~~~~~~~~
vendor/intel/external/android_ia/hwcomposer/os/android/drmhwctwo.cpp:590:15: error: use of non-static data member 'is_explicit_sync_enabled_' of 'DrmHwcTwo' from nested type 'HwcDisplay'
          if (is_explicit_sync_enabled_) {
              ^~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.

Comment (txu4) [inline os/android/drmhwctwo.h]: @kalyankondapally the explicit sync query is done in hwc initialization, when the common display objects, like render are not created yet. we cannot assign its member variables. The current static is a workaround. I prefer to let it in first. Idealy, we need refactoring the code, and use a singleton class to manage the drm feature and capability.

Comment (kalyankondapally) [inline os/android/drmhwctwo.h]: what we could do is pass the state to display, display queue whom should be already creaed at this point. Once displayqueue starts to initialize compositor etc it can set that info appropriately

Comment (mwang30) [inline os/android/drmhwctwo.h]: 1. As I mentioned:
renderer_ need to be initialized in Compositor::Init(). If not, below call would caused fault address error.
renderer_->SetExplicitSync(ExplicitSyncEnabled);
So if don't touch Compositor::Init(), we should use static is_explicit_sync_enabled_ in GLRenderer and EGLOffScreenContext.

2. DrmHwcTwo and DrmHwcTwo::HwcDisplay both use is_explicit_sync_enabled_ in os/android/drmhwctwo.cpp. If not use static, there is compile issue mentioned above.

**Comment by kalyankondapally** [comment]
https://github.com/01org/IA-Hardware-Composer/pull/59/commits/f8077163073ff4b1a10ac66bfb4547a0e58a503f
https://github.com/01org/IA-Hardware-Composer/pull/59/commits/7067df41252c418816b69246afbe69c96318e47a
https://github.com/01org/IA-Hardware-Composer/pull/59/commits/0a5f55a2ed425c8f4c8254efc0536129aa40883d

Should simplify this support. Now, what you can do is just pass the value to DisplayQueue, and if disabling explicit sync support is true you can mark out_fence_ptr_prop_ = 0, if false,   GetDrmObjectProperty("OUT_FENCE_PTR", crtc_props, &out_fence_ptr_prop_); and see if the kernel supports it. Rest should take care of it self. Can you try with that series of patches in that pull request + your changes and see. I think we would need to make another release branch from TOT after this for AT also with https://github.com/01org/IA-Hardware-Composer/pull/58 included

**Comment by kalyankondapally** [comment]
merged thanks
